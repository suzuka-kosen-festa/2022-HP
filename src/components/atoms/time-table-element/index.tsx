import type { FC } from "react";
import tw from "twin.macro";
import React, { useMemo, useRef, useState } from "react";
import dayjs from "dayjs";
import type { TimeTableElementProperties } from "./types/model";
import { useKeyPress } from "../../../hooks/useKeyPress";
import { useClickAway } from "../../../hooks/useClickAway";
import { Portal } from "../portal";
import Icon from "../icon";

const TimeTableElementBox = tw.div`flex flex-col cursor-pointer px-1 py-0.5 sm:(px-6 py-4)`;

const Overlay = tw.div`flex fixed justify-center items-center inset-0 z-50 overflow-y-auto bg-gray-300 bg-opacity-70`;

const ModalWrapper = tw.section`flex items-center bg-white border-primary-yellow px-5.5 py-6 border-3 rounded-3xl w-72 md:(px-21 py-16 border-5 rounded-5xl w-180)`;

const ModalBox = tw.div`relative w-full max-w-lg mx-auto text-center space-y-4 space-x-4`;

const TimeTableElement: FC<TimeTableElementProperties> = ({ event }) => {
  const [isShow, setShow] = useState<boolean>(false);
  const ref = useRef<HTMLDivElement>(null);
  // eslint-disable-next-line consistent-return
  const backgroundColor = useMemo(() => {
    switch (event.stage) {
      case "main":
        return tw`bg-primary-red`;
      case "sub":
        return tw`bg-primary-blue`;
      case "live":
        return tw`bg-primary-purple`;
      case "game":
        return tw`bg-primary-green`;
      // no default
    }
  }, [event.stage]);
  const times = useMemo(() => {
    const start = dayjs(event.start_time).format("HH:mm");
    const end = dayjs(event.end_time).format("HH:mm");
    return { start, end };
  }, [event.end_time, event.start_time]);
  useClickAway(ref, () => setShow(false));
  useKeyPress("Escape", () => setShow(false));
  return (
    <>
      <TimeTableElementBox css={backgroundColor} onClick={() => setShow(true)}>
        <time css={tw`font-zen font-bold text-white text-xss md:text-xl select-none`}>
          <span css={tw`after:content-["~"]`}>{times.start}</span>
          {times.end}
        </time>
        <p css={tw`font-zen font-bold text-white text-xss md:text-xl select-none`}>{event.title}</p>
      </TimeTableElementBox>
      <Portal portalId="#modal">
        {isShow ? (
          <Overlay>
            <ModalWrapper ref={ref}>
              <ModalBox>
                <Icon
                  as="button"
                  icon="charm:cross"
                  css={tw`absolute -top-14 right-0 md:(-top-20 -right-12) text-primary-yellow`}
                  tabIndex={0}
                  onClick={() => setShow(false)}
                />
                <h2 css={tw`font-zen font-medium text-text text-base md:text-3.5xl select-none`}>{event.title}</h2>
                <p css={tw`font-zen font-medium text-text text-xs md:text-2xl select-none`}>{event.descriptions}</p>
              </ModalBox>
            </ModalWrapper>
          </Overlay>
        ) : null}
      </Portal>
    </>
  );
};

export { TimeTableElement };
