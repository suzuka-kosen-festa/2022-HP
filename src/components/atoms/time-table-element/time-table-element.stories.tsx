import type { ComponentStoryObj, ComponentMeta } from "@storybook/react";
import { ComponentPropsWithoutRef } from "react";
import { TimeTableElement } from ".";

type T = typeof TimeTableElement;
type Story = ComponentStoryObj<T>;
type Meta = ComponentMeta<T>;

const args: ComponentPropsWithoutRef<T> = {
  title: "オープニング",
  descriptions: "空白の2年間、たくさんの協力を得てついに開催！溜め込んだものを全て出し切れ！高専祭開幕です！",
  start_time: "2022-10-30 9:00",
  end_time: "2022-10-30 9:30",
  stage: "main",
};

export default {
  args,
  argTypes: {
    title: {
      description: "Title of event",
      control: {
        type: "text",
      },
    },
    descriptions: {
      description: "Description of event",
      control: {
        type: "text",
      },
    },
    start_time: {
      description: "Start time of event",
      control: {
        type: "text",
      },
    },
    end_time: {
      description: "End time of event",
      control: {
        type: "text",
      },
    },
    stage: {
      description: "Stage category of event",
      control: {
        type: "select",
        options: ["main", "sub", "live", "game"],
      },
    },
  },
  component: TimeTableElement,
} as Meta;

export const Default: Story = {};
export const Main: Story = {
  args: {
    stage: "main",
  },
};
export const Sub: Story = {
  args: {
    stage: "sub",
  },
};
export const Live: Story = {
  args: {
    stage: "live",
  },
};
export const Game: Story = {
  args: {
    stage: "game",
  },
};
