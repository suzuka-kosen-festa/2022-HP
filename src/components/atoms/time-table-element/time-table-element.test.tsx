import { composeStories } from "@storybook/testing-react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import * as stories from "./time-table-element.stories";

const { Default } = composeStories(stories);

describe("(components) atoms/time-table-element", () => {
  beforeAll(() => {
    const portalRoot = document.createElement("div");
    portalRoot.setAttribute("id", "modal");
    document.body.appendChild(portalRoot);
  });
  test("to be atoms", () => {
    const { container } = render(<Default />);
    expect(container).toBeAtom();
  });
  test("take snap shot", () => {
    const { container } = render(<Default />);
    expect(container).toMatchSnapshot();
  });
});
