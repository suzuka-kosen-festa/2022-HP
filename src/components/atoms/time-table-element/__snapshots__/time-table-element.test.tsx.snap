// Vitest Snapshot v1

exports[`(components) atoms/time-table-element > take snap shot 1`] = `
<div>
  <li
    class="css-1melwrt-TimeTableElement"
  >
    <time
      class="css-cj0c87-TimeTableElement"
    >
      <span
        class="css-zu2x60-TimeTableElement"
      >
        09:00
      </span>
      09:30
    </time>
    <p
      class="css-1e4vuyv-TimeTableElement"
    >
      オープニング
    </p>
  </li>
</div>
`;
